stdlib/public/core/Slab.swift

```diff
@@ -0,0 +1,499 @@
+  ///   `startIndex`.
+  /// - Returns: The index value immediately before `i`.
+  @available(SwiftStdlib 6.1, *)
+  @_alwaysEmitIntoClient
+  @_transparent
+  public borrowing func index(before i: Int) -> Int {
+    i &- 1
+  }
+
+  /// Accesses the element at the specified position.
+  ///
+  /// The following example accesses an element of an array through its
+  /// subscript to print its value:
+  ///
+  ///     var streets = ["Adams", "Bryant", "Channing", "Douglas", "Evarts"]
+  ///     print(streets[1])
+  ///     // Prints "Bryant"
+  ///
+  /// You can subscript a collection with any valid index other than the
+  /// collection's end index. The end index refers to the position one past
+  /// the last element of a collection, so it doesn't correspond with an
+  /// element.
+  ///
+  /// - Parameter position: The position of the element to access. `position`
+  ///   must be a valid index of the collection that is not equal to the
+  ///   `endIndex` property.
+  ///
+  /// - Complexity: O(1)
+  @available(SwiftStdlib 6.1, *)
+  @_addressableSelf
+  @_alwaysEmitIntoClient
+  public subscript(_ i: Int) -> Element {
+    @_transparent
+    unsafeAddress {
+      _precondition(indices.contains(i), "Index out of bounds")
+
+      return _address + i
+    }
+
+    @_transparent
+    unsafeMutableAddress {
+      _precondition(indices.contains(i), "Index out of bounds")
+
+      return _mutableAddress + i
+    }
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Reduce and Swap
+//===----------------------------------------------------------------------===//
+
+@available(SwiftStdlib 6.1, *)
+extension Slab where Element: ~Copyable {
+  /// Exchanges the values at the specified indices of the vector.
+  ///
+  /// Both parameters must be valid indices of the vector and not
+  /// equal to `endIndex`. Passing the same index as both `i` and `j` has no
+  /// effect.
+  ///
+  /// - Parameters:
+  ///   - i: The index of the first value to swap.
+  ///   - j: The index of the second value to swap.
+  ///
+  /// - Complexity: O(1)
+  @available(SwiftStdlib 6.1, *)
+  @_alwaysEmitIntoClient
+  public mutating func swapAt(
+    _ i: Int,
+    _ j: Int
+  ) {
+    guard i != j else {
+      return
+    }
+
+    _precondition(indices.contains(i), "Index out of bounds")
+    _precondition(indices.contains(j), "Index out of bounds")
+
+    let ithElement = _mutableBuffer.moveElement(from: i)
+    let jthElement = _mutableBuffer.moveElement(from: j)
+    _mutableBuffer.initializeElement(at: i, to: jthElement)
+    _mutableBuffer.initializeElement(at: j, to: ithElement)
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Span
+//===----------------------------------------------------------------------===//
+
+@available(SwiftStdlib 6.1, *)
+extension Slab where Element: ~Copyable {
+
+  @available(SwiftStdlib 6.1, *)
+  public var storage: Span<Element> {
```

Comment:
Needs `@_addressableSelf`, `@_alwaysEmitIntoClient`, and preferably `@_transparent`.